File: src/JsonServices.Core/Events/ClientSubscriptionManager.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace JsonServices.Events
{
	public class ClientSubscriptionManager
	{
		private ConcurrentDictionary<string, ConcurrentDictionary<string, IClientSubscription>> Subscriptions { get; } =
			new ConcurrentDictionary<string, ConcurrentDictionary<string, IClientSubscription>>();

		public Action Add<TEventArgs>(ClientSubscription<TEventArgs> sub)
			where TEventArgs : EventArgs
		{
			var subs = Subscriptions.GetOrAdd(sub.EventName, n => new ConcurrentDictionary<string, IClientSubscription>());
			subs[sub.SubscriptionId] = sub;
			return () => subs.TryRemove(sub.SubscriptionId, out var ignored);
		}

		public void Broadcast(string eventName, EventArgs eventArgs)
		{
			if (Subscriptions.TryGetValue(eventName, out var list))
			{
				foreach (var sub in list.Values)
				{
					sub.Invoke(this, eventArgs);
				}
			}
		}

		public void BroadcastAsync(string eventName, EventArgs eventArgs)
		{
			ThreadPool.QueueUserWorkItem(x =>
			{
				try
				{
					Broadcast(eventName, eventArgs);
				}
				catch
				{
					// TODO: handle exceptions
				}
			});
		}
	}
}

File: sample/sample-client/src/JsonServices/ClientSubscriptionManager.ts
import { EventEmitter } from 'eventemitter3';
import { ClientSubscription } from './ClientSubscription';

export class ClientSubscriptionManager {
    private emitter = new EventEmitter();
    private subscriptions: { [subscriptionId: string]: ClientSubscription; } = {};

    public add = (subscription: ClientSubscription) => {
        this.subscriptions[subscription.subscriptionId] = subscription;
        this.emitter.on(subscription.eventName, subscription.invoke, subscription);

        return () => {
            delete this.subscriptions[subscription.subscriptionId];
            this.emitter.off(subscription.eventName, subscription.invoke, subscription);
        }
    }

    public broadcast = (eventName: string, eventArgs: object) => {
        this.emitter.emit(eventName, eventArgs);
    }
}